Add new payment processor (ednx-pp).

vad33r opened the pull request:
This PR Adds a new payment processor based on Payu, the main objective create a flexible payment processor.

Documentation of the payment processor: [ednx-pp](https://docs.google.com/document/d/13OZufrorVAXO4O51P1Q-8qaJMSsraEuAdLFsb-R0Xkw/edit?usp=sharing)

andrey-canon commented:
This pr is too long, you can split it, test are missing

andrey-canon commented on ecommerce_extensions/payment/sandbox_ednxPP/pp_sandbox/.vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
?

vad33r commented:
@andrey-canon should i split it in 2 PR, one for the payment processor and the other for the sandbox? what do you think?

vad33r committed:
Adding the edupay payment processor.

Includes all the following commits by @vad3r
---

Base of ednx payment processor

Remove api key from parameters dictionary to avoid sending it in the POST request

Updating ecommerce_extensions/urls.py and ecommerce_extensions/payment/urls.py

Fixing processor and view of ednx_pp

Import settings for transactions codes

Change name of ednxpp to edupay

Adding tests for edupay

Fixing logs of edupay view

Making tests pass for edupay

Updating name of urls

Adding dict for extra parameters and class for payment status

Updating status view

Add docstring a class property for name_pp

Updating code when payment is pending

Set Name as a configurable attribute

Updating msg in template pending_status and add new url to view status of payment

Allow configurable parameters with user and basket.

Adding hash in status view

Add new template for payment pending status

Use configurable keys in signature process.

just accept str parameters from basket and user

Add encryption utils

Updating template for payment status and fixing bug when status is declined

Fixing template of status

Add constants.py file and fixing edupay view

Split post function of EdnxPaymentResponseView in three methods

Add cryptography and waffle to requirements

Add django-crum to requirements

Fixing tests

Fix pylint tests

Fix isort tests

Updating class name for payment processor

Rebase

Change name of variable to avoid line too long

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+""" ednx payment processing. """
```
please improve this docstring

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+class ednx_pp(BasePaymentProcessor):
+    """
+    ednx payment processor
+
```
improve docstring and remove blank space

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+SourceType = get_model('payment', 'SourceType')
+
+
+class ednx_pp(BasePaymentProcessor):
```
the class name should be camel case, and please improve that name

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+
+    """
+
+    NAME = 'ednx_pp'
```
improve this name, could this be configurable ?

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+
+        Raises:
+            KeyError: If no settings configured for this payment processor
+            AttributeError: If LANGUAGE_CODE setting is not set.
```
where are you using the language code ?

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        self.payment_page_url = configuration['payment_page_url']
+        self.merchant_id = configuration['merchant_id']
+        self.account_id = configuration['account_id']
+        self.api_key = configuration['api_key']
```
this comment was made before your last commit , ignore it

vad33r replied:
I define apikey here to use it in the creation of the signature

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+    """
+
+    NAME = 'ednx_pp'
+    TRANSACTION_ACCEPTED = '4'
```
could the transaction codes be configurable ?  maybe someone uses different codes

vad33r replied:
Yes, this is just for tests, but we can define differente codes.

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            'payment_page_url': self.payment_page_url,
+            'merchantId': self.merchant_id,
+            'accountId': self.account_id,
+            'ApiKey': self.api_key,
```
same comment as above

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        Returns:
+            dict: Base parameters required to complete a transaction, including a signature.
+        """
+        parameters = {
```
I think these are the fixed parameters and there will be optional parameters, so could you rename this variable ?

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            'accountId': self.account_id,
+            'ApiKey': self.api_key,
+            'referenceCode': basket.order_number,
+            'tax': self.tax,
```
IMO this could be optional

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            'ApiKey': self.api_key,
+            'referenceCode': basket.order_number,
+            'tax': self.tax,
+            'taxReturnBase': self.tax_return_base,
```
optional ...

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            GatewayError: Indicates a general error on the part of the processor.
+            InvalidPayUDecision: Indicates an unknown decision value
+        """
+
```
unnecessary blank space

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        # Raise an exception for payments that were not accepted. Consuming code should be responsible for handling
+        # and logging the exception.
+        transaction_state = response['state_pol']
+        if transaction_state != self.TRANSACTION_ACCEPTED:
```
where it the right place to handle the pending state ?

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        if transaction_state != self.TRANSACTION_ACCEPTED:
+            exception = {
+                self.TRANSACTION_DECLINED: TransactionDeclined,
+                self.TRANSACTION_ERROR: GatewayError
```
trailing comma

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        Raises:
+            TransactionDeclined: Indicates the payment was declined by the processor.
+            GatewayError: Indicates a general error on the part of the processor.
+            InvalidPayUDecision: Indicates an unknown decision value
```
returns is missing

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        the parameters that they are giving to us.
+
+        Arguments:
+            parameters (dict): A dictionary of transaction parameters.
```
signature_type is missing

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        # signatures to validate confirmation response
+        if signature_type == self.CONFIRMATION_SIGNATURE:
+            value = parameters['value']
+            last_decimal = value[-1]
```
blank space after this line

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            if last_decimal == '0':
+                new_value = value[:-1]
+            else:
+                new_value = value
```
```suggestion
            new_value = value[:-1] if last_decimal == '0' else value
```

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+            else:
+                new_value = value
+
+            uncoded = "{api_key}~{merchant_id}~{reference_sale}~{new_value}~{currency}~{state_pol}".format(
```
this block of code is very similar to the previous one, I think there is a way to do this shorter

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        to accept payment refunds
+        see http://developers.payulatam.com/en/api/refunds.html
+        """
+
```
unnecessary blank space

andrey-canon commented on ecommerce_extensions/payment/processors/ednx_pp.py:
```diff
@@ -0,0 +1,238 @@
+        """
+        Return the first product encountered in the basket with the product
+        class of 'seat'.  Return None if no such products were found.
+        """
```
arguments and returns statements are missing

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+""" ednx payment processing views """
```
improve doc-string

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+        return super().dispatch(request, *args, **kwargs)
+
+    def _get_billing_address(self, ednx_response):
+
```
missing doc string and unnecessary blank space

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+        # Note (CCB): Orders should not be created until the payment processor has validated the response's signature.
+        # This validation is performed in the handle_payment method. After that method succeeds, the response can be
+        # safely assumed to have originated from ednx.
+        ednx_response = request.POST.dict()
```
does it work with request.data ?

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+            logger.info(
+                'Received PP merchant notification for transaction [%s], associated with basket [%d].',
+                transaction_id,
+                basket_id
```
trailing comma

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+            basket_id = OrderNumberGenerator().basket_id(order_number)
+
+            logger.info(
+                'Received PP merchant notification for transaction [%s], associated with basket [%d].',
```
PP ?

vad33r replied:
PP = Payment Processor

andrey-canon replied:
there more than 3 payment processor, include the name or something like that

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+                    self.handle_payment(ednx_response, basket)
+                except InvalidSignatureError:
+                    logger.exception(
+                        'Received an invalid ednx response. The payment response was recorded in entry [%d].',
```
invalid ednx response ?

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+                        'The payment response was recorded in entry [%d].',
+                        basket.id,
+                        exception.__class__.__name__,
+                        ppr.id
```
trailing comma

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+                    logger.exception(
+                        'PP payment failed for basket [%d]. The payment response was recorded in entry [%d].',
+                        basket.id,
+                        ppr.id
```
trailing comma

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+                request=request,
+            )
+
+            return HttpResponse()
```
what is the status here ? could you be explicit ?

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+        except (ValueError, ObjectDoesNotExist):
+            return None
+
+    def post(self, request):
```
there are three try-except block, in order to improve the readability you can slip them in three private methods

andrey-canon commented on ecommerce_extensions/payment/views/ednx_pp.py:
```diff
@@ -0,0 +1,209 @@
+            site_configuration=basket.site.siteconfiguration,
+        )
+
+        try:
```
this block doesn't make sense, since this catches the errors from `return redirect(receipt_url)` and then make the same  in the line 209 without any change

andrey-canon reviewed (changes requested):
the urls file is missing and tests are missing

andrey-canon commented:
tests are still failing, since you change the files name a lot of comments have not been addressed

andrey-canon commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,240 @@
+    ednx payment processor
+    """
+    NAME = 'edupay'
+    NAME_PP = 'Poli'
```
you shouldn't hard-code this

vad33r replied:
this is just for test

andrey-canon commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,240 @@
+    """
+    NAME = 'edupay'
+    NAME_PP = 'Poli'
+    TRANSACTION_ACCEPTED = getattr(settings, "accepted_code", "4")
```
does this make sense, if we can't change this in runtime ?

andrey-canon commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,240 @@
+        Returns:
+            dict: Base parameters required to complete a transaction, including a signature.
+        """
+        parameters = {
```
you have not addressed  previous comments

andrey-canon commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,240 @@
+            card_type=card_type,
+        )
+
+    def _generate_signature(self, parameters, signature_type):
```
this block doesn't consider a default value so this will fail when signature_type is not PAYMENT_FORM_SIGNATURE or  CONFIRMATION_SIGNATURE

andrey-canon commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,240 @@
+        This method should be implemented in the future in order
+        to accept payment refunds
+        """
+        logger.exception(
```
is this longer than 120 characters ?

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+
+
+class EdnxPaymentResponseView(EdxOrderPlacementMixin, View):
+    """ Validates a response from PP and processes the associated basket/order appropriately. """
```
Add examples  and more details for this view

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+        return super().dispatch(request, *args, **kwargs)
+
+    def _get_billing_address(self, edupay_response):
+
```
docstring

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+        )
+
+    def _get_basket(self, basket_id):
+        if not basket_id:
```
docstring

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+            return None
+
+    def post(self, request):
+        """Process a PP merchant notification and place an order for paid products as appropriate."""
```
Again  don't use `PP`

andrey-canon replied:
I think this is a previous comment, this method is too long , split it

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+                        basket.id,
+                        ppr.id,
+                    )
+                    basket.status = 'PENDING'
```
is this a valid status ?

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+                    )
+                    basket.status = 'DECLINED'
+                    basket.save()
+                    return HttpResponse()
```
what is the code ?

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+                        basket.id,
+                        ppr.id,
+                    )
+                    return HttpResponse()
```
same comment as above

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+        # pylint: disable=unused-argument
+        """Handle an incoming user returned to us by PP after processing payment."""
+        edupay_response = request.GET.dict()
+        try:
+            transaction_id = edupay_response.get('transactionId')
+            order_number = edupay_response.get('referenceCode')
+            basket_id = OrderNumberGenerator().basket_id(order_number)
+
+            logger.info(
+                'Received [%s] payer notification for transaction [%s], associated with basket [%d].',
+                EdnxPP.NAME,
+                transaction_id,
+                basket_id,
+            )
+
+            basket = self._get_basket(basket_id)
+
+            if not basket:
```
duplicate

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+        )
+
+        try:
+            return redirect(receipt_url)
```
non-sense block , execute redirect, catch the exception and then execute redirect again

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+        else:
+            try:
+                order = Order.objects.get(number=order_id)
+            except:  # pylint: disable=bare-except
```
This raises and specific exception, why are you catching everything

felipemontoya replied:
What do you suggest instead? catch `Order.DoesNotExist` ?

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+
+    def _get_payment_status(self, order_number):
+        """Get the current state of the payment."""
+        check_order = order_number
```
why did you assign this to a different variable  ?

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+            basket_id = OrderNumberGenerator().basket_id(check_order)
+            basket = Basket.objects.get(id=basket_id)
+            if basket.status == 'PENDING':
+                logger.info(
```
same log with a different word

andrey-canon commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,296 @@
+                return 'complete'
+        except:  # pylint: disable=bare-except
+            pass
+
```
You don't log when there is an error ...

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+{% endif %}
+
+{% comment %}
+    TODO: Remove once this has a MFE, and asking the lms if there is a discount happens in it
```
what ?

vad33r replied:
solved, extending the base file.

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+
+{% if msg == "pending" %}
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "Esperando la confirmación de pago." as tmsg %}{{ tmsg | force_escape }}</h2>
```
English is the default language

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "Esperando la confirmación de pago." as tmsg %}{{ tmsg | force_escape }}</h2>
+        <div class="confirm-message">
+            Gracias, su orden estara lista cuando el pago sea confirmado.
```
English is the default language

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    </div>
+{% elif msg == "closed" %}
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "Su pago ha sido rechazado." as tmsg %}{{ tmsg | force_escape }}</h2>
```
English is the default language

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "Su pago ha sido rechazado." as tmsg %}{{ tmsg | force_escape }}</h2>
+        <div class="confirm-message">
+            Hubo un problema a la hora de validar su pago y no pudo concretarse.
```
English is the default language

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    </div>
+{% elif msg == "error" %}
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "No se encontro ningun pago registrado." as tmsg %}{{ tmsg | force_escape }}</h2>
```
English is the default language

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    <div class="receipt container content-container">
+        <h2 class="thank-you">{% trans "No se encontro ningun pago registrado." as tmsg %}{{ tmsg | force_escape }}</h2>
+        <div class="confirm-message">
+            La orden que busca no existe.
```
English is the default language, if there is not order this should return 404

vad33r replied:
solved

andrey-canon commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,112 @@
+    </div>
+{% endif %}
+
+{% compress js %}
```
?

vad33r replied:
solved

mariajgrimaldi commented on ecommerce_extensions/payment/encryption_utils.py:
```diff
@@ -0,0 +1,39 @@
+    """
+    encryption_key = getattr(settings, 'PAYMENT_ENCRYPTION_KEY', 'encryption_key')
+    encryption_key = encryption_key.encode()  # Convert to type bytes
+    kdf = PBKDF2HMAC(
```
what is kdf?

vad33r replied:
@andrey-canon

felipemontoya replied:
kdf is short for Key Derivation Function, commonly used named in cryptography.

PB - KDF - 2HMAC

mariajgrimaldi commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,384 @@
+DEFAULT_NAME = 'edupay'
+
+
+class EdnxPP(BasePaymentProcessor):
```
Please use more descriptive names

vad33r replied:
EdnxPaymentProcessor? what do you think?

mariajgrimaldi commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,384 @@
+            value = parameters['value']
+            parameters['value'] = value[:-1] if value[-1] == '0' else value
+            signature_keys = self.ednx_configuration.get('SIGNATURE_CONFIRMATION_KEYS', [])
+
```
what happens when is `else`?

vad33r replied:
what do you mean exactly? when signature type isn't PAYMENT_FORM_SIGNATURE or CONFIRMATION_SIGNATURE?

felipemontoya replied:
Yes, that is exactly the unresolved case here. What happens then?

felipemontoya replied:
Talking to @andrey-canon is clear now. If the type is not accounted for here then in line 354 uncoded will be = to basic_uncoded which is desirable and normal in some configurations

mariajgrimaldi commented on ecommerce_extensions/payment/tests/views/tests_edupay.py:
```diff
@@ -0,0 +1,46 @@
+from ecommerce_extensions.payment.views.edupay import EdnxPaymentResponseView
+
+
+class EdnxPPViewTest(TestCase):
```
please use more descriptive names

vad33r replied:
EdnxPaymentProcessorViewTest?

mariajgrimaldi replied:
That's better!

mariajgrimaldi commented on ecommerce_extensions/payment/tests/views/tests_edupay.py:
```diff
@@ -0,0 +1,46 @@
+        response = EdnxPaymentResponseView.as_view()(request)
+
+        self.assertEqual(status.HTTP_400_BAD_REQUEST, response.status_code)
+        order_generator_mock().basket_id.assert_called_once_with(data["reference_sale"])
```
just one test?

vad33r replied:
i removed the others tests beacause they use a get method that not exist anymore, and i try to used it in the status view but isn't work, so, ir order to make the tests pass i decide to work in fix pylint and isort tests.

felipemontoya replied:
Having just one test is very low test quality. At this point we might need to continue with this one test only, but it is almost the same as no testing

mariajgrimaldi commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,276 @@
+                return HttpResponse(status=status.HTTP_400_BAD_REQUEST)
+        finally:
+            # Store the response in the database.
+            ppr = self.payment_processor.record_processor_response(
```
more descriptive names, please

vad33r replied:
oks

mariajgrimaldi commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,276 @@
+        """Provide confirmation of payment."""
+        order_id = decode_string(order_number)
+        payment_status = self._get_payment_status(order_id)
+        if payment_status != 'Submitted':
```
why not `payment_status.lower() != 'submitted'` just in case

mariajgrimaldi commented:
Remember to resolve conflicts

felipemontoya commented on ecommerce_extensions/payment/encryption_utils.py:
```diff
@@ -0,0 +1,39 @@
+""" Moduel to encript the order number in the url for status page """
```
```suggestion
""" Module to encrypt the order number in the URL for status page """
```

felipemontoya commented on ecommerce_extensions/payment/urls.py:
```diff
@@ -1,12 +1,17 @@
 from django.urls import path
 
-from ecommerce_extensions.payment.views import fomopay, payu
+from ecommerce_extensions.payment.views import edupay, fomopay, payu
```
```suggestion
from ecommerce_extensions.payment.views import (
    edupay,
    fomopay,
    payu,
)
```

felipemontoya commented on ecommerce_extensions/payment/urls.py:
```diff
@@ -1,12 +1,17 @@
 """Payment processors urls file."""
```
I find it weird that we change the file permissions. I will change this locally and squash push the file with the same permissions as the others

felipemontoya commented on ecommerce_extensions/templates/payment/pending_status.html:
```diff
@@ -0,0 +1,34 @@
+        </div>
+    {% elif msg == "error" %}
+        <div class="receipt container content-container">
+            <h2 class="thank-you">{% trans "Error 404" as tmsg %}{{ tmsg | force_escape }}</h2>
```
Is this seriously a 404 error?  How can we tell this is actually a page not found instead of something else.

felipemontoya commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,384 @@
+        This method should be implemented in the future in order
+        to accept payment refunds
+        """
+        logger.exception(
+            'edupay processor can not issue credits or refunds',
+        )
```
```suggestion
        logger.exception( 'edupay processor can not issue credits or refunds')
```

felipemontoya commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,276 @@
+        InvalidSignatureError
+    )
+
+    class EdxOrderPlacementMixin:
```
We should move this to a test_shims.py file somewhere

felipemontoya replied:
Also, do we need the "get_receipt_page_url" function in the tests?

felipemontoya replied:
I'll leave it here in this cycle

felipemontoya commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,276 @@
+        else:
+            try:
+                order = Order.objects.get(number=order_id)  # pylint: disable=unused-variable
+            except:  # pylint: disable=bare-except
```
```suggestion
            except Order.DoesNotExist:
```

felipemontoya commented on ecommerce_extensions/payment/views/edupay.py:
```diff
@@ -0,0 +1,276 @@
+                '[%s] Payment: Basket not found',
+                EdnxPaymentProcessor.NAME,
+            )
+
+        return 'error'
```
```suggestion

        logger.info(
            '[%s] Payment: Unknown processing error',
            EdnxPaymentProcessor.NAME,
        )
        return 'error'
```

felipemontoya reviewed (changes requested):
I'm trying to get this over the line where we can push this to Prod.

@andrey-canon I resolved the issues you had raised that where ready and left some others as suggestions. Can we review this during a sync call and push it over the line?

@vad33r I left some questions that I need you to answer please. I need to understand what was your intent

felipemontoya commented on ecommerce_extensions/payment/processors/edupay.py:
```diff
@@ -0,0 +1,384 @@
+        """
+        Returns True if this payment processor is enabled, and False otherwise.
+        """
+        return waffle.switch_is_active('{}{}'.format(
```
Can we change this to a setting so that we remove the Waffle dependencies

felipemontoya committed:
Addressing the PR feedback:
----

Apply suggestions from code review

Fixing the permissions of urls.py files

Moving exceptions to its own file

Removing waffle in favor of a regular setting from the current settings dict

felipemontoya reviewed (approved):
From my side we are ready here. I fixed errors, applied suggestions and squashed.
CI tests are passing now. @andrey-canon if you try this locally and it works, we are ready to merge
